### ISSUE TITLE
optimization algorithm for searching best tuning parameters

### ISSUE THREAD

**Issue by rladeira** [topepo/caret#321]
Hello,

Firstly, I would like to congratulate everyone involved in the development of caret package. This package is just amazing!

I was wondering why it doesn't implement some others algorithms for searching for optimal tuning parameters.

What would be the caveats of using a genetic algorithm, for instance, instead of grid or random search? Do you think using some of those powerful optimization algorithms for tuning parameters is a good idea?

**Comment by elephann** [comment]
Currently, caret employed grid search, which is universal for all models, and random search which has limits for some models. Besides above two, Max adopted adaptive resampling that may speed up the choice of optimal tune parameter,involve "gls" and "BT" two approach. 
However, the two adaptive resampling methods may result in quite different tuning parameters, which  is the optimal is not quite sure. 
The great challenge, I think, still the choice of optimal tuning parameter. For toy data, any approach above mentioned may take several hours even with parallel running. While for the real world data this may take several days. Based on "no free lunch" principle, nearly three hundred model in caret, trial and error may quite time consuming or even impossible.
So, I do agree with rladeira, not only for the best parameter selection but also with the efficiency consideration.

**Comment by topepo** [comment]
Basically the issue with using a search procedure is related to how you compute the fitness value. The main method to get a decent estimate (caveats below) that avoids overfitting is to resample. Resampling multiples the computational costs and most non-gadient based search routines use _a lot_ of function evaluations. Another issue is stopping. Many of these routines will spend a lot of function evaluations on fine tuning the final answer to achieve convergence. To me, that isn't time well spent alt although you could probably tweak the convergence criteria to fix this. 

For the fitness value, you could hold out a validation set like they did in the olden days of neural networks but a lot of models will quickly overfit to these data. However, if the validation set is large, there is less to worry about and this would probably be a good strategy.

Another option for some models would be the OOB performance estimate, although I do think that you would overfit there eventually.    

If you think that AIC or related values are good, then that is a fast alternative. I don't think that they are directly related to model efficacy as other metrics though.

**Comment by rladeira** [comment]
I understand your point, Max, the convergence issue is really a key observation. I haven't thought about that. However, as these convergence issues may lead the algorithms to spend a long time trapped too close of an excellent solution, a reasonably sized grid search usually spends a lot of time away of the optimum zone. Maybe a mixed approach could do better, do you know anything of this sort in the literature?

Just one more question: Given that resampling (using 5 10-folds cv, for example) is affordable, do you think a powerful optimization algorithm would still likely overfit the data?

**Comment by topepo** [comment]
I took a little time to write some code and test out GA's. Here is an example below and I'll do a blog post about this (prob on Monday)...

```r
library(caret)
set.seed(17516)
training_data <- SLC14_1(500)
testing_data <- SLC14_1(10^5)
 
svm_fit <- function(x) {
  mod <- train(y ~ ., data = training_data,
               method = "svmRadial",
               preProc = c("center", "scale"),
               trControl = trainControl(method = "cv"),
               tuneGrid = data.frame(C = 2^x[1], sigma = exp(x[2])))
  -getTrainPerf(mod)[, "TrainRMSE"]
}

gbm_fit <- function(x) {
  mod <- train(y ~ ., data = training_data,
               method = "gbm",
               trControl = trainControl(method = "cv", number = 10),
               tuneGrid = data.frame(n.trees = floor(x[1])+1, 
                                     interaction.depth = floor(x[2])+1,
                                     shrinkage = x[3],
                                     n.minobsinnode = floor(x[4])+1),
               verbose = FALSE)
  -getTrainPerf(mod)[, "TrainRMSE"]
}
library(GA)
library(kernlab)
library(gbm)
svm_ga_obj <- ga(type = "real-valued", 
                 fitness = svm_fit,
                 ## test cost values between ~0 and 2^10,
                 ## test sigma values between exp(-5) and 1
                 min = c(-5, -5), 
                 max = c(10, 0), 
                 popSize = 50, 
                 maxiter = 12,
                 seed = 16478,
                 keepBest = TRUE,
                 monitor = NULL,
                 elitism = 2)

gbm_ga_obj <- ga(type = "real-valued", 
                 fitness = gbm_fit,
                 ## Trees between [1, 5000], 
                 ## depth between [1, 11]
                 ## shrinkage between [0, 0.2]
                 ## terminal node size in [5, 25]
                 min = c(   1,  1, 0, 5), 
                 max = c(5000, 11, .2, 25), 
                 popSize = 50, 
                 maxiter = 12,
                 seed = 513,
                 keepBest = TRUE,
                 monitor = NULL)
```

**Comment by rladeira** [comment]
Really nice piece of code! Thanks for sharing and taking your time to address my question! 

Just one thing came to my mind. By returning the value produced by -getTrainPerf(mod)[, "TrainRMSE"], you are optimizing the aggregated RMSE across resamples in validation or training set?  According to how the search procedures in caret works, i suppose it is the validation set, but since the naming is somewhat misleading here, I'm asking to be sure.

I'll try to produce a similar version using the diferential evolution algorithm, which is considered by some people one of the best options for real valued optimization.

**Comment by rladeira** [comment]
Using your code as a template, it was very easy to apply the differential evolution.

```
library(caret)
library(parallel)
library(doMC)

set.seed(17516)
training_data <- SLC14_1(500)
testing_data <- SLC14_1(10^5)

registerDoMC(cores = detectCores())

svm_fit <- function(x) {
  mod <- train(y ~ ., data = training_data,
               method = "svmRadial",
               preProc = c("center", "scale"),
               trControl = trainControl(method = "cv"),
               tuneGrid = data.frame(C = 2^x[1], sigma = exp(x[2])))
  getTrainPerf(mod)[, "TrainRMSE"]
}

library(DEoptim)
library(kernlab)

svm_de_obj <-  DEoptim(fn = svm_fit,
                       ## test cost values between ~0 and 2^10,
                       ## test sigma values between exp(-5) and 1
                       lower = c(-5, -5), 
                       upper = c(10, 0),
                       control = DEoptim.control(reltol = 1e-3,
                                                                steptol = 10,
                                                                itermax = 100))

fitted_params <- svm_de_obj$optim$bestmem

svm_model <- train(y ~ ., data = training_data,
                   method = "svmRadial",
                   preProc = c("center", "scale"),
                   trControl = trainControl(method = "cv", number = 10),
                   tuneGrid = data.frame(C = 2^fitted_params[1], sigma = exp(fitted_params[2])))

predictions <- predict(svm_model, testing_data)

cat("Validation RMSE:", getTrainPerf(svm_model)[, "TrainRMSE"], "\n")
cat("Test RMSE:", RMSE(predictions, testing_data$y))


```

**Comment by rladeira** [comment]
For the sake of curiosity, I executed the GA version with 31 iterations and the same seed to compare. The result was:

Validation RMSE: 5.866318 
Test RMSE: 6.029585

One remarkable difference is that the default pop size for DEoptim is 15, while the GA was executed with the 50 as popsize (exactly the same setting in your code).

Since this post is just a question, not really an issue, I'll reopen it, so that is more likely that more people can read and contribute to this discussion.

**Comment by tobigithub** [comment]
Interesting, some general comments, 
1) Using svmPoly with grid parameters  (degree ,C , scale) performs even better with the DEoptim code example (Global Optimization by Differential Evolution) from above. Convergence after 14 iterations (names should be training/testing, I just use the same style as above).

Validation RMSE: 5.271413 
Test RMSE: 4.989915

2) Other models may not even need grid optimization and give a decent result by themselves, see below sorted by RMSE (no GA tuning). Of course in this case the only two methods that match the GA optimized  method = "svmRadial" are svmPoly and  gaussprPoly. These results were not tested for over-fitting.

````R
Num   Name        R^2     RMSE    time [s] Model name
1     svmPoly     0.915   5.349   13.960   Support Vector Machines with Polynomial Kernel
2     gaussprPoly 0.905   5.852   20.940   Gaussian Process with Polynomial Kernel
3     rvmPoly     0.801   7.445   50.900   Relevance Vector Machines with Polynomial Kernel
4     xgbTree     0.667   9.309   6.000    eXtreme Gradient Boosting
5     xgbLinear   0.624   9.519   82.340   eXtreme Gradient Boosting
6     cubist      0.704   9.612   14.060   Cubist
7     krlsRadial  0.669   9.682   46.250   Radial Basis Function Kernel Reg LS
8     extraTrees  0.652   9.737   56.830   Random Forest by Randomization
````
2) the SLC14_1 dataset is synthetic and has a small variable number,
in such as case genetic algorithms itself (instead of SVMs) maybe faster and accurate. 

From the [caret documentation](http://www.inside-r.org/node/192179) The function SLC14_1 simulates a system from Sapp et al. (2014). All informative predictors are independent Gaussian random variables with mean zero and a variance of 9. The prediction equation is:
 x_1 + sin(x_2) + log(abs(x_3)) + x_4^2 + x_5*x_6 +  I(x_7*x_8*x_9 < 0) + I(x_10 > 0) + x_11*I(x_11 > 0) +  sqrt(abs(x_12)) + cos(x_13) + 2*x_14 + abs(x_15) +  I(x_16 < -1) + x_17*I(x_17 < -1) - 2 * x_18 - x_19*x_20 
The random error here is also Gaussian with mean zero and a variance of 9.

3) For example an external genetic algorithm would come up after a few seconds/minutes calculation with an much better Test RMSE of 3.96 and R^2=0.96 plus you would directly get an approximation function or maybe even the correct function (which is hard as the above has if/then functionality). 

4) I agree with the above comment by Max, that unless you have unlimited computer/cluster resources there needs to be a specific design to make sure that the optimization time is spent efficiently and not  in some costly fine-tuning with minor improvements.

5) It would be  nice if that would work for all caret models and grid tunings in general, plus it would be important to have some performance annotations, so one could estimate that the calculation will take hours/days or longer.  Here the simulation functions such as  SLC14 would be of course useful again.

**Comment by topepo** [comment]
@rladeira I didn't think to use `DEoptim`. That seems to work very well. If you could describe the optimization process a bit (and contrast it to GAs), would you like to write a guest post about this on the blog? I'll probably add one for the GA results tonight/tomorrow.

**Comment by rladeira** [comment]
@topepo, this would be really nice. Yes, I do have interest, it's a nice opportunity to investigate further the performance of differential evolution in this context. However, I think I will need a couple weeks to start the writing, because I'm really out of time in the next weeks. Is it ok?

**Comment by topepo** [comment]
That would be fine with me. I've not be on top of blog posts this year.

The only thing that I'd ask is that you use the Rmd format since it makes
it easier for me to convert it for the blog. I also have some code (not
shown on the website) that you can embed in the `svm_fit` call that will
track how the search occurs in detail.

Drop me an email at `max.kuhn@pfizer.com` and we can go over any details
(whenever you get to it).

Thanks,

Max
